File: ext/src/Function.c
/*
  Copyright (c) 2017 DataStax, Inc.

  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

#include "php_driver.h"

zend_class_entry *php_driver_function_ce = NULL;

ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
ZEND_END_ARG_INFO()

static zend_function_entry php_driver_function_methods[] = {
  PHP_ABSTRACT_ME(Function, name, arginfo_none)
  PHP_ABSTRACT_ME(Function, simpleName, arginfo_none)
  PHP_ABSTRACT_ME(Function, arguments, arginfo_none)
  PHP_ABSTRACT_ME(Function, returnType, arginfo_none)
  PHP_ABSTRACT_ME(Function, signature, arginfo_none)
  PHP_ABSTRACT_ME(Function, language, arginfo_none)
  PHP_ABSTRACT_ME(Function, body, arginfo_none)
  PHP_ABSTRACT_ME(Function, isCalledOnNullInput, arginfo_none)
  PHP_FE_END
};

void php_driver_define_Function(TSRMLS_D)
{
  zend_class_entry ce;

  INIT_CLASS_ENTRY(ce, PHP_DRIVER_NAMESPACE "\\Function", php_driver_function_methods);
  php_driver_function_ce = zend_register_internal_class(&ce TSRMLS_CC);
  php_driver_function_ce->ce_flags |= ZEND_ACC_INTERFACE;
}

File: ext/doc/Dse/Function.php
<?php

/*
  Copyright (c) 2016 DataStax, Inc.
  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

namespace Dse;

/**
 * A PHP representation of a function
 */
interface Function_ {

    /**
     * Returns the full name of the function
     * @return string Full name of the function including name and types
     */
    public function name();

    /**
     * Returns the simple name of the function
     * @return string Simple name of the function
     */
    public function simpleName();

    /**
     * Returns the arguments of the function
     * @return array Arguments of the function
     */
    public function arguments();

    /**
     * Returns the return type of the function
     * @return Dse\Type Return type of the function
     */
    public function returnType();

    /**
     * Returns the signature of the function
     * @return string Signature of the function (same as name())
     */
    public function signature();

    /**
     * Returns the lanuage of the function
     * @return string Language used by the function
     */
    public function language();

    /**
     * Returns the body of the function
     * @return string Body of the function
     */
    public function body();

    /**
     * Determines if a funciton is called when the value is null.
     * @return bool Returns whether the function is called when the input columns are null
     */
    public function isCalledOnNullInput();

}
